### DISCUSSION TITLE
Generating values for related fields

### DISCUSSION THREAD

**Discussion by foaw** [instancio/instancio#866] [Q&A]
The POJO I'm instancio-ating is a resource that can be created, modified and removed subsequently. In doing so, it stores the instants of when that was done and by whom. There are a few of simple rules, however:

- The time of creation as well as the author must be set for all resources: `(createdAt != null) && (createdBy != null)`
- A resource cannot be modified before it is created: `modifiedAt >= createdAt`
- The instant a resource is removed is also the last time it's modified, as it becomes locked for further modifications: `(removedAt != null) -> (modifiedAt == removedAt)`

<details>
  <summary>Resource definition for a better understanding</summary>

```java
public record Resource
    (Instant createdAt,
     String createdBy,
     Instant modifiedAt,
     String modifiedBy,
     Instant removedAt)
{}
```
</details>

Bearing this in mind, I want every second instance to be modified, and every sixth of those to be removed. The problem arises where I want the creation instant be set necessarily, and the modification time to depend on both `createdAt` and `modifiedBy`.

Before introduction of `createdAt`, my approach was to generate `modifiedBy`, choosing one of `null`, or `"admin"`. Then based on the value of `modifiedBy`—should it be non-null—I assigned a custom stateful generator. On the first call it returns a random instant for `modifiedAt`, stores the returned value, and on the subsequent call returns that stored value with a probability of 1/6.

Now I also want to restrict `JavaTimeTemporalGenerator#min` to only get instants after `createdAt`, so as not to set a `modifiedAt` before `createdAt`, but I can't access the actual `createdAt`.

I was thinking along the lines of writing my own `InstantGenerator`, which would store the last returned `createdAt` in something like `AtomicReference`, and refer to that in the generator of `modifiedAt`/`removedAt`. However, with this approach I would probably have to test my tests. 😅

**Comment by armandino** [comment]
> I was thinking along the lines of writing my own InstantGenerator, which would store the last returned createdAt in something like AtomicReference, and refer to that in the generator of modifiedAt/removedAt. However, with this approach I would probably have to test my tests. 😅

There were a couple of similar discussions recently, but your use case is somewhat more complicated :) I'm not sure what you described can be easily achieved. It'll require some experimenting and might end up being too complicated to be worth it. I think your best bet is to construct the objects manually.

Just in case, below are the other posts. I'll follow up if I can think of a good approach.

- #856
- #843

**Reply by foaw** [reply]
This is but upsetting. Do you think:

- The entire context could be provided to generators, including those values already evaluated, instead of or in addition to `Random`? However, this would mean quite an undertaking of refactoring to `InstancioEngine#createObject` and related areas.

- We can have a method where the origin value from `Assign#given` would be passed to a user-defined `Function<Origin, Destination>`, like `Assign.given(Resource::createdAt).derive(When.isNotNull(), createdAt -> ...)`? This only covers the use-cases where only a single value is needed. Ideally, then `Assign#given` should be enhanced to handle `SelectorGroup`'s, like `all(field(Resource::createdAt), field(Resource::modifiedBy))`.

**Comment by foaw** [comment]
As for my particular case, I think `createdAt` could be generated in isolation from `modifiedAt` and `removedAt`. The instant generators would be restrained in such a way that `modifiedAt`'s `TemporalGeneratorSpec#range` would only allow values before `createdAt`, and `createdAt`'s—only values after `modifiedAt`, splitting the timeline in half.

I'm leaving the discussion open for now, and I will continue to search for a solution less convoluted.

**Reply by foaw** [reply]
I actually just changed my approach altogether: if a POJO is so sophisticated, maybe it's worth just writing a custom generator for the whole POJO. That is, not generators for individual properties, but generate the whole POJO inside of one custom generator. All the options you have with `Model` you can extract as configuration methods like in `GeneratorSpec`'s.

If you think about it, you will later, one way or another, find yourself occasionally needing to deviate from the standard `Model`; speaking of `Resource`'s, you may want all resources generated to have `removedAt`, which isn't an easy goal to accomplish with `Model`'s.
